### ISSUE TITLE
Provide standard a (de-)serializer for Kotlin's Regex class

### ISSUE THREAD

**Issue by sschuberth** [Kotlin/kotlinx.serialization#887]
**What is your use-case and why do you need this feature?**

I'd like to (de-)serialize `kotlin.text.Regex` as part of a `data class`.

**Describe the solution you'd like**

With `kotlin.text.Regex` being a Kotlin class, it would be nice if (de-)serialization worked out of the box.

**Comment by ypawar159** [comment]
I would also like this implemented.
It's better to compile the regex at deserilaization and use that object multiple times than compiling the regex each time required.

I have tried the following but didn't work. Learning kotlin coming from python so don't know what is wrong here.

```kotlin
object RegexSerializer : KSerializer<Regex> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Regex", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: Regex) {
        encoder.encodeString(value.pattern)
    }

    override fun deserialize(decoder: Decoder): Regex {
        return Regex(decoder.decodeString())
    }
}
```

**Comment by sschuberth** [comment]
I was using the same code [here](https://github.com/sschuberth/stan/blob/cb0ddd1656e58382371be3f41aa00bbe6ee74938/lib/src/main/kotlin/utils/RegexSerializer.kt), and it actually worked for me.

**Comment by ypawar159** [comment]
Thanks for the reference code.
I haven't got my code working yet.
Have you used it for nested property?

**Comment by sschuberth** [comment]
Not sure how "nested" it should be, but I've used it for files like [this one](https://github.com/sschuberth/stan/blob/cb0ddd1656e58382371be3f41aa00bbe6ee74938/lib/src/main/resources/config.json#L15).

**Comment by sandwwraith** [comment]
@ypawar159 Have you applied your serializer to `Regex` property as in the guide: https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/serializers.md#specifying-serializer-on-a-property ?

**Comment by ypawar159** [comment]
I was using it for a complete class with reference from other part of the documentation.
Will try with this approach as well and post results here.

**Comment by ypawar159** [comment]
@sandwwraith Annotating the specific field in the data class worked successfully. Thanks for the reference.
Then I tried to do the same for the `List<Regex>` field but it fails to compile. Anything specific to be done for the list of it?

**Comment by sandwwraith** [comment]
@ypawar159 Try to use annotation on the type itself: `List<@Serializable(RegexSerializer::class) Regex>`

Or just `@file:UseSerializers(RegexSerializer::class)` if you have a lot of them.

**Comment by ypawar159** [comment]
Annotating the field with `List<@Serializable(RegexSerializer::class) Regex>` is working as expected.
I also tried to use the same for Regex field as well. `val url_regex: @Serializable(RegexSerializer::class) Regex` and this also seems to be working fine.
Is there any difference using it this way rather than the following?
```kotlin
@Serializable(with = RegexSerializer::class)
val url_regex: Regex
```

**Comment by ypawar159** [comment]
How about we add this info in the documentation or in the FAQ section?

**Comment by sschuberth** [comment]
@qwwdfsad are you planning to add a built-in Regex (de-)serializer as you assigned this to yourself?

**Comment by sandwwraith** [comment]
Closing as won't fix, discussion here: https://github.com/Kotlin/kotlinx.serialization/pull/2137#issuecomment-1384137980

(Regex serializer still may be implemented as a part of #1931)
